consensus/XDPoS/engines/engine_v2/engine.go

```diff
@@ -496,22 +515,124 @@ func (x *XDPoS_v2) UpdateMasternodes(chain consensus.ChainReader, header *types.
 }
 
 func (x *XDPoS_v2) VerifyHeader(chain consensus.ChainReader, header *types.Header, fullVerify bool) error {
-	return nil
+	err := x.verifyHeader(chain, header, nil, fullVerify)
+	if err != nil {
+		log.Warn("[VerifyHeader] Fail to verify header", "fullVerify", fullVerify, "blockNum", header.Number, "blockHash", header.Hash(), "error", err)
+	}
+	return err
 }
 
-// TODO: Yet to be implemented XIN-135
+// Verify a list of headers
 func (x *XDPoS_v2) VerifyHeaders(chain consensus.ChainReader, headers []*types.Header, fullVerifies []bool, abort <-chan struct{}, results chan<- error) {
 	go func() {
-		for range headers {
+		for i, header := range headers {
+			err := x.verifyHeader(chain, header, headers[:i], fullVerifies[i])
+			log.Warn("[VerifyHeaders] Fail to verify header", "fullVerify", fullVerifies[i], "blockNum", header.Number, "blockHash", header.Hash(), "error", err)
 			select {
 			case <-abort:
 				return
-			case results <- nil:
+			case results <- err:
 			}
 		}
 	}()
 }
 
+// Verify individual header
+func (x *XDPoS_v2) verifyHeader(chain consensus.ChainReader, header *types.Header, parents []*types.Header, fullVerify bool) error {
+	// If we're running a engine faking, accept any block as valid
+	if x.config.V2.SkipV2Validation {
+		return nil
+	}
+	_, check := x.verifiedHeaders.Get(header.Hash())
+	if check {
+		return nil
+	}
+
+	if header.Number == nil {
+		return utils.ErrUnknownBlock
+	}
+
+	if fullVerify {
+		if len(header.Validator) == 0 {
+			return consensus.ErrNoValidatorSignature
+		}
+		// Don't waste time checking blocks from the future
+		if header.Time.Int64() > time.Now().Unix() {
+			return consensus.ErrFutureBlock
+		}
+	}
+
+	// Verify this is truely a v2 block first
+	var decodedExtraField utils.ExtraFields_v2
+	err := utils.DecodeBytesExtraFields(header.Extra, &decodedExtraField)
+	if err != nil {
+		return utils.ErrInvalidV2Extra
+	}
+	quorumCert := decodedExtraField.QuorumCert
+	if quorumCert == nil || quorumCert.Signatures == nil || len(quorumCert.Signatures) == 0 {
+		return utils.ErrInvalidQC
+	}
+
+	if quorumCert.ProposedBlockInfo.Hash == (common.Hash{}) {
+		return utils.ErrEmptyBlockInfoHash
+	}
+
+	// Nonces must be 0x00..0 or 0xff..f, zeroes enforced on checkpoints
+	if !bytes.Equal(header.Nonce[:], utils.NonceAuthVote) && !bytes.Equal(header.Nonce[:], utils.NonceDropVote) {
+		return utils.ErrInvalidVote
+	}
+	// Ensure that the mix digest is zero as we don't have fork protection currently
+	if header.MixDigest != (common.Hash{}) {
+		return utils.ErrInvalidMixDigest
+	}
+	// Ensure that the block doesn't contain any uncles which are meaningless in XDPoS_v1
+	if header.UncleHash != utils.UncleHash {
+		return utils.ErrInvalidUncleHash
+	}
+
+	// Verify v2 block that is on the epoch switch
+	if header.Validators != nil {
```

Comment:
If `IsEpochSwitch()`, we go to this branch.
If `!IsEpochSwitch`, then `len(header.Validators) == 0 `should be true